_posts/2024-07-01-zio-fullstack-webapp.md

```diff
@@ -0,0 +1,844 @@
+  def makeZIO: URIO[PersonService, PersonController] = // (3)
+    ZIO.service[PersonService].map(new PersonController(_))
+}
+
+```
+
+* **(1)** Dependency Injection.
+* **(2)** Controller implementation, where json codec are transparently handled.
+* **(3)** Service instantiation:  URIO[PersonService, PersonController]
+
+
+Everything is IO, URIO[PersonService, PersonController] is an IO that needs a PersonService and will produce a PersonController and cannot fail.
+
+> **URIO** is a type alias for a ZIO that cannot fail:
+>
+>   type **URIO**[R, A] = ZIO[R, **Nothing**, A]`
+
+**zServerLogic** is part of the Tapir ZIO support, and bridges a ZIO in the Rest API.
+
+**zServerLogic** is a combinator that will transform a ZIO effect into a server logic, it will handle the error and the success case.
+
+
+### 6.3. Client side
+
+On **client side**, Tapir will allow to interact as client, il will leverage the same endpoint definition to generate the client side code.
+
+It will smartly generate the client side code, that will be used in the Laminar UI.
+
+
+## 7. Laminar
+
+Laminar is a 100% Scala reactive UI library. In few word, Laminar provide reactive variable:
+
+* **Signal**, to watch an unique element
+* **EventStream**, to consume stream of events
+* **Var**: to watch and update an element
+* **EventBus**: to consumes and produces events.
+
+Laminar also provide Dom manipulation facilities, and bindings to WebComponent libraries (UI5 in this example).
+
+
+## 7.1. ZIO, Tapir and Laminar
+
+On **client side**, Tapir will allow to interact as client, is a very straigh forward way: 
+
+```scala
+ Button(
+  "-- Post -->",
+  onClick --> { _ => // (1)
+    PersonEndpoint // (2)
+      .createEndpoint(personVar.now()) // (2a andThen 2b)
+      .emitTo(userBus) // (3a and 3b)
+  }
+)
+```
+
+
+* **(1)** from an Laminar UI event
+* **(2)** a HTTP endpoint
+  * **(2a)** reads the current value of a Laminar Var
+  * **(2b)** turns into ZIO effect to handle the HTTP request
+* **(3)** this effect
+  * **(3a)** this effect will run
+  * **(3B)** its result is consumed by the Laminar UI
+
+
+Whoah, this is a lot of magic in a single line.
+
+In the same way that on server side **zServerLogic** bridges ZIO in Tapir through an extension method, on client side two extension methods are used:
+* From laminar to ZIO
+* From ZIO to Laminar back.
+
+This is the power of Scala3 extension methods, that allow to extend existing classes with new methods.
+
+A first extension, turn this endpoint value to a function ( def **apply …** ):
+
+```scala
+extension [I, E <: Throwable, O](endpoint: Endpoint[Unit, I, E, O, Any])
+    /**
+     * Call the endpoint with a payload, and get a ZIO back.
+     * @param payload
+     * @return
+     */
+    def apply(payload: I): RIO[BackendClient, O] =
+      ZIO.service[BackendClient]
+        .flatMap(_.endpointRequestZIO(endpoint)(payload))
+```
+
+Function, when called that turns this pimped endpoint to a function that:
```

Comment:
Which function?
Also let's avoid the old "pimping" terms from the days of implicit classes